New Post: CSS `size-adjust` (#5523)
GoogleChrome/web.dev: src/site/content/en/blog/css-size-adjust/index.md

hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/fL3dWSY7YBfagFP0ikMW.jpg
alt: Well used letters from a letterpress, set into rows
tags:
  - blog
  - css
date: 2021-06-04
updated: 2021-06-04
---

Consider the following
[demo](https://codepen.io/argyleink/pen/8ace843980e2fd24bf15c8bd1bd072e4?editors=1100)
where the `font-size` is a consistent `64px`, and the only difference between each of these headers is the `font-family`. The examples on the left have not been adjusted and
have an inconsistent final size. The examples on the right use `size-adjust` to
ensure `64px` is the consistent final size.

<figure class="w-figure w-figure--fullbleed">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/6xhatUhszn7M6pDFcVgV.mp4",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">
   In this example Chrome DevTools CSS grid layout debug tools are used to show heights.
  </figcaption>
</figure>

This post explores a [new CSS font-face
descriptor](https://drafts.csswg.org/css-fonts-5/#size-adjust-desc) called

> `size-adjust`, available in Firefox 89 (behind a flag) and Chromium 92 (see

Reviewer: I suggest listing Chromium first, since that's the shipping one.

Author: ```suggestion
`size-adjust`, available in Chromium 92 and Firefox 89 (behind a flag); (see
```

[caniuse](https://caniuse.com/mdn-css_at-rules_font-face_size-adjust) for latest
support). It also demonstrates a few ways to correct and normalize font sizes
for smoother user experience, consistent design systems and more predictable

> page layout. For example, when optimizing initial font rendering to prevent

Reviewer: ```suggestion
page layout. One important use case is optimizing web font rendering to prevent
```

[cumulative layout shift](/cls/) (CLS). 

Here's an interactive [demo](https://codepen.io/argyleink/pen/rNyMjxR) of the
problem space. Try changing the typeface with the dropdown and observe: 

> 1. The wildheight differences in the results.

Reviewer: ```suggestion
1. The height differences in the results.
```

> 1. Visual jarring content shifts. 

Reviewer: ```suggestion
1. Visually jarring content shifts. 
```

1. Moving interactive target areas (the dropdown jumps around!).

{% Codepen {
  user: 'web-dot-dev',
  id: 'WNpzRKd'
} %}

{% Aside 'key-term' %} 
Font family vs Typeface: A **typeface** is referred to by
its family name plus its font face. `Helvetica Bold` is referring to the
specific bold typeface while `Helvetica` is generically referring to the entire

> family of 8+. With CSS `@font-face` you'll be dealing with typefaces, even

Reviewer: ```suggestion
family of 8+ typefaces (normal, bold, italic, etc). With CSS `@font-face` you'll be dealing with typefaces, even
```

> though to use them you need to write `font-family`. 

Reviewer: Suggest removing this "even though", or rephrasing to make clear that it makes sense because there are multiple fields in an `@font-face` that *together* form the parameters indicating which typeface to use.

[demo](https://codepen.io/argyleink/pen/8ace843980e2fd24bf15c8bd1bd072e4?editors=1100)
where the only difference between each of these headers is the `font-family`,
same `font-size` of `64px`. The examples on the left have not been adjusted and
have an inconsistent final size. The examples on the right use `size-adjust` to
ensure `64px` is the consistent final size.

<figure class="w-figure w-figure--fullbleed">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/6xhatUhszn7M6pDFcVgV.mp4",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">
    Chrome DevTools CSS grid layout debug tools used to show heights
  </figcaption>
</figure>

This post explores a [new CSS font-face
descriptor](https://drafts.csswg.org/css-fonts-5/#size-adjust-desc) called
`size-adjust`, available in Firefox 89 (behind a flag) and Chromium 92 (see
[caniuse](https://caniuse.com/mdn-css_at-rules_font-face_size-adjust) for latest
support). It also demonstrates a few ways to correct and normalize font sizes
for smoother user experience, consistent design systems and more predictable
page layout. For example, when optimizing initial font rendering to prevent

> [cumulative layout shift](https://web.dev/cls/) (CLS). 

Reviewer 2: ```suggestion
[cumulative layout shift](/cls/) (CLS). 
```

{% endAside %}

## How to scale fonts with `size-adjust`

An introduction to the syntax:

```css/2
@font-face {
  font-family: "Adjusted Typeface";
  size-adjust: 150%;
  src: url(some/path/to/typeface.woff2) format('woff2');
}
```

1. Creates a custom typeface named `Adjusted Typeface`.
1. Uses `size-adjust` to scale up each glyph to 150% their default size.
1. Affects only the single imported typeface.

Use the above custom typeface like this:

```css
h1 {
  font-family: "Adjusted Typeface";
}
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'yLMKgRx'
} %}

{% Aside 'warning' %} 
If the second headline in the above demo is not larger than the
first one, your browser does not support `size-adjust`. 
{% endAside %}

### Mitigating CLS with seamless font swapping
In the following gif, watch the examples on the left and how there's a shift
when the font is changed. This is just a small example with a single headline
element and the issue is very noticeable.

<figure class="w-figure">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/6xhatUhszn7M6pDFcVgV.mp4",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">

>     left has layout shift, right does not

Reviewer 2: ```suggestion
    The example on the left has layout shift, the one on the right does not.
```

  </figcaption>
</figure>

> To improve font rendering, a great technique is [font
> swapping](https://web.dev/font-display/), where you render a quick-loading
> system font to show the content first, and then swap that with a web font when
> the web font finishes loading. This gives you the best of both worlds: the
> content is visible as soon as possible, and you get a nicely styled page without
> sacrificing your user's time to content. The problem, however, is that sometimes
> when the web font loads, it shifts the entire page around since it presents at a
> slightly different box height size, even though the font is set to the same
> size.

Reviewer 2: ```suggestion
To improve font rendering, a great technique is [font
swapping](/font-display/). Render a quick-loading
system font to show the content first, then swap that with a web font when
the web font finishes loading. This gives you the best of both worlds: the
content is visible as soon as possible, and you get a nicely styled page without
sacrificing your user's time to content. The problem however, is that sometimes
when the web font loads, it shifts the entire page around since it presents at a
slightly different box height size, even though the font is set to the same
size.
```

<figure class="w-figure w-screenshot">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/F7CwbZuMHaapGTeXRfSJ.mp4",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">
    more content equals more potential layout shift when font swaps
  </figcaption>
</figure>

By putting `size-adjust` in the `@font-face` rule, it sets a global glyph
adjustment for the font face. Timing this right will lead to minimal visual
change, a seamless swap. To create a seamless swap, hand adjust or try this
[size-adjust
calculator](https://deploy-preview-15--upbeat-shirley-608546.netlify.app/perfect-ish-font-fallback)
by [Malte Ubl](https://twitter.com/cramforce).

<figure class="w-figure w-screenshot">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/Tw8J2e0ymyGogeA7NJjY.mp4",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">

>     choose a [Google Web Font](https://fonts.google.com/), get back a pre-adjusted @font-face snippet

Reviewer 2: ```suggestion
    Choose a <a href="https://fonts.google.com/">Google Web Font</a>, get back a pre-adjusted <code>@font-face</code> snippet.
```

  </figcaption>
</figure>

At the beginning of this post, fixing the font size issue was done by trial and
error. `size-adjust` was nudged in the source code until the same header in
`Cookie` and `Arial` rendered the same `64px` as `Press Start 2P` did naturally.

> I aligned 2 fonts to a target font size.

Reviewer 2: ```suggestion
I aligned two fonts to a target font size.
```

  font-family: 'Adjusted Arial';
  size-adjust: 86%; 
  src: local(Arial);
}

@font-face {
  font-family: 'Cookie';
  size-adjust: 90.25%;
  src: url(...woff2) format('woff2');
}
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'jOBzyop'
} %}

### Calibrating fonts

You as the author determine the calibration target(s) for normalizing font
scale. You might normalize on Times, Arial, or a system font, then adjust custom
fonts to match. Or, you might adjust local fonts to match what you download. 

Strategies for `size-adjust` calibration:
1. **Remote target:**<br>Adjust local fonts towards downloaded fonts.
1. **Local target:**<br>Adjust downloaded fonts towards local target fonts.

#### Example 1: Remote target

Consider the following snippet which adjusts a locally available font to size
match a remote src custom font. A remote custom font is the target for
calibration, a brand font perhaps:

```css/2-3
@font-face {
  font-family: "Adjusted Regular Arial For Brand";
  src: local(Arial);
  size-adjust: 90%;
}

@font-face {
  font-family: "Rad Brand";
  src: url(some/path/to/a.woff2) format('woff2');
}

html {
  font-family: "Rad Brand", "Adjusted Regular Arial For Brand";
}
```

> Local regular Arial adjusting in anticipation of a loaded custom font, for a
> seamless swap. 

Reviewer 2: ```suggestion
In this example, local font Arial is adjusting in anticipation of a loaded custom font, for a
seamless swap. 
```

The order of `font-family` is critical. It's where order
and priority go. Ensure the typeface you want the most, is first. Furthermore,
`local(Arial)` may not be available on all your user's devices, therefore it's
important to provide multiple fallbacks in the font-family. 
{% endAside %}

To create a fully cross platform adjustment, see the following example which provides 2 adjusted local fallback fonts in anticipation of a brand font.

```css
@font-face {
    font-family: "Roboto-fallback-1";
    size-adjust: 100.06%;
    src: local("Arial");
}

@font-face {
    font-family: "Roboto-fallback-2";
    size-adjust: 99.94%;
    src: local("Arimo");
}

@font-face {
  font-family: "Roboto Regular";
  src: url(some/path/to/roboto.woff2) format('woff2');
}

html {
  font-family: "Roboto Regular", "Roboto-fallback-1", "Roboto-fallback-2";
}
```

#### Example 2: Local target

Consider the following snippet which adjusts a brand custom font to match Arial:

```css/3
@font-face {
  font-family: "Rad Brand - Adjusted For Arial";
  src: url(some/path/to/a.woff2) format('woff2');
  size-adjust: 110%;
}

html {
  font-family: "Rad Brand - Adjusted For Arial", Arial;
}
```

> Adjusting incoming downloaded fonts to a local target for a seamless swap.

Reviewer 2: ```suggestion
```

alt="scissors above and blow the word overrides, demonstrating the areas the
features can trim to", width="800", height="136" %}

### `ascent-override` 

The `ascent-override` descriptor defines the height above the baseline of a
font.

```css/3
@font-face {
  font-family: "Ascent Adjusted Arial Bold";
  src: local(Arial Bold);
  ascent-override: 71%;
}
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'ZEexebZ'
} %}

The red headline (unadjusted) has space above it's capital letter A and O, while
the blue headline has been adjusted so it's [cap
height](https://en.wikipedia.org/wiki/Cap_height) fits snug against the overall
bounding box.

### `descent-override`

The `descent-override` descriptor defines the height below the baseline of the
font.

```css/3
@font-face {
  font-family: "Ascent Adjusted Arial Bold";
  src: local(Arial Bold);
  descent-override: 0%;
}
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'OJpvpNQ'
} %}

The red headline (unadjusted) has space below it's D and O
[baselines](https://en.wikipedia.org/wiki/Baseline_(typography)), while the blue
headline has been adjusted so it's letters rest snug on the baseline.

### `line-gap-override`

>  The `line-gap-override` descriptor defines the line-gap metric for the font.
>  This is the font recommended line-gap or external leading

Reviewer 2: ```suggestion
 The `line-gap-override` descriptor defines the line-gap metric for the font.
 This is the font recommended line-gap or external leading.
```
